support/controlplane-component/status.go

```diff
@@ -0,0 +1,167 @@
+package controlplanecomponent
+
+import (
+	"fmt"
+	"strings"
+
+	hyperv1 "github.com/openshift/hypershift/api/hypershift/v1beta1"
+	assets "github.com/openshift/hypershift/control-plane-operator/controllers/hostedcontrolplane/v2/assets"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/utils/ptr"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (c *controlPlaneWorkload) checkDependencies(cpContext ControlPlaneContext) ([]string, error) {
+	if len(c.dependencies) == 0 {
+		return nil, nil
+	}
+
+	componentsList := &hyperv1.ControlPlaneComponentList{}
+	if err := cpContext.Client.List(cpContext, componentsList, client.InNamespace(cpContext.HCP.Namespace)); err != nil {
+		return nil, err
+	}
+
+	unavailableDependencies := sets.New(c.dependencies...)
+	for _, component := range componentsList.Items {
+		if !unavailableDependencies.Has(component.Name) {
+			continue
+		}
+
+		availableCondition := meta.FindStatusCondition(component.Status.Conditions, string(hyperv1.ControlPlaneComponentAvailable))
+		if availableCondition != nil && availableCondition.Status == metav1.ConditionTrue {
+			unavailableDependencies.Delete(component.Name)
+		}
+	}
+
+	return sets.List(unavailableDependencies), nil
+}
+
+func (c *controlPlaneWorkload) reconcileComponentStatus(cpContext ControlPlaneContext, component *hyperv1.ControlPlaneComponent, unavailableDependencies []string, reconcilationError error) error {
+	component.Status.Resources = []hyperv1.ComponentResource{}
+	if err := assets.ForEachManifest(c.Name(), func(manifestName string) error {
+		adapter, exist := c.manifestsAdapters[manifestName]
+		if exist && adapter.predicate != nil && !adapter.predicate(cpContext) {
+			return nil
+		}
+
+		obj, gvk, err := assets.LoadManifest(c.name, manifestName)
+		if err != nil {
+			return err
+		}
+
+		component.Status.Resources = append(component.Status.Resources, hyperv1.ComponentResource{
+			Kind:  gvk.Kind,
+			Group: gvk.Group,
+			Name:  obj.GetName(),
+		})
+
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	if len(unavailableDependencies) > 0 {
+		meta.SetStatusCondition(&component.Status.Conditions, metav1.Condition{
+			Type:    string(hyperv1.ControlPlaneComponentAvailable),
```

Comment:
maybe we should use a different condition here (unless we only want to use dependencies for initial deploy). During an upgrade, a component can be available but still be waiting for dependencies. It might be useful to mimic the ocp clusteroperator conditions (available, progressing, degraded). It is also possible for the component to be available and to have failed reconciliation.

Reply:
the way I see it, is that a component's functionality requires its dependency to be available, so during an upgrade, a dependency upgrading means its not available and therefore the dependent component functionality is hindered so we mark it as unavailable (although technically its deployment is available).

e.g. if etcd is upgrading, then KAS can't be operational during that time and we mark it as unavailable until etcd finishes the upgrade and kas can start upgrading.